File: 2_Periodo/AED_II/AED_Estudo/TADL/Fila.java
package TADL;

public class Fila {
    int[] array;
    int primeiro, ultimo;

    Fila(){
        this(5);
    }

    Fila(int tamanho){
    array = new int[tamanho + 1];
    primeiro = ultimo = 0;
    }

    void inserir(int x) throws Exception{
        if(((ultimo + 1) % array.length) == primeiro)
            throw new Exception("erro");
        
        array[ultimo] = x;
        ultimo = (ultimo + 1) % array.length;
    }

    int remover() throws Exception{
        if(primeiro == ultimo){
            throw new Exception("erro");
        }

        int removido = array[primeiro];
        primeiro = (primeiro + 1) % array.length;
        return removido;
    }

    void mostrar(){
        int i = primeiro;

        while((i % array.length) != ultimo){
            System.out.println(array[i]);
            i = (i + 1) % array.length;
        }
    }
}

File: 2_Periodo/AED_II/AED_Estudo/TADL/fila.c
int array[MAXTAM + 1];
int primeiro, ultimo;

void start(){
    primeiro = ultimo = 0;
}

void inserir(int x){
    if(((ultimo + 1) % (MAXTAM + 1)) == primeiro){
        exit(1);
    }

    array[ultimo] = x;
    ultimo = (ultimo + 1) % (MAXTAM + 1);
}

int remover(){
    if(primeiro == ultimo){
        exit(1);
    }

    int removido = array[primeiro];
    primeiro = (primeiro + 1) % MAXTAM;
}

